### PR TITLE
client: hold the client_lock in C_Readahead::finish

### PR DESCRIPTION
**Pull request by lxbsz**
Fixes: https://tracker.ceph.com/issues/49225
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by jtlayton — APPROVED** [review]
Looks correct, and quite necessary. Note that without this, I believe the callback could race with the get_cap_ref in _read_async.

**Comment by lxbsz** [comment]
jenkins test api

**Comment by lxbsz** [comment]
> Looks correct, and quite necessary. Note that without this, I believe the callback could race with the get_cap_ref in _read_async.

Yeah, right. 
I just raised one tracker https://tracker.ceph.com/issues/49225 if we need to backport it.
Thanks.

**Comment by vshankar** [comment]
jenkins test api

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
libcephfs tests pass: https://pulpito.ceph.com/vshankar-2021-02-15_13:40:18-fs-wip-vshankar-testing-20210215-050616-distro-basic-smithi/

I'm also running basic fuse related test for this PR, but the results have not come back yet.

**Comment by lxbsz** [comment]
https://pulpito.ceph.com/xiubli-2021-02-22_04:53:19-fs:functional-wip-lxb-testing-20210221-1527-distro-basic-smithi

There has two jobs dead and all the others succeeded. @vshankar @batrick

**Review by batrick — APPROVED** [review]

File: src/client/Client.cc [hunk 1]
```diff
@@ -9762,6 +9762,7 @@ Client::C_Readahead::~C_Readahead() {
 }
 
 void Client::C_Readahead::finish(int r) {
+  std::scoped_lock cl(client->client_lock);
```
Comment (batrick) [inline src/client/Client.cc:9765]: So actually, the object cacher has the `client_lock` already and locks it when finishing these contexts. See `ObjectCacher::bh_read_finish`. This change cause a deadlock (due to recursive locks):

/ceph/teuthology-archive/pdonnell-2021-02-25_20:20:39-fs-wip-pdonnell-testing-20210225.184709-distro-basic-gibba/5913567/teuthology.log

````
(gdb) bt
#0  0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffc03d06d8) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7fffc03d0688, cond=0x7fffc03d06b0) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x7fffc03d06b0, mutex=0x7fffc03d0688) at pthread_cond_wait.c:655
#3  0x00007f40dc0f58bc in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007f40e5553213 in std::condition_variable::wait<C_SaferCond::wait()::{lambda()#1}>(std::unique_lock<std::mutex>&, C_SaferCond::wait()::{lambda()#1}) (__p=...,
    __lock=..., this=0x7fffc03d06b0) at /usr/include/c++/7/condition_variable:99
#5  C_SaferCond::wait (this=0x7fffc03d0680) at ./src/common/Cond.h:100
#6  0x00007f40e54f18df in Client::_read_async (this=this@entry=0x1c80cd0, f=f@entry=0x1d6d8e0, off=off@entry=131072, len=len@entry=131072, bl=bl@entry=0x7fffc03d08b0)
    at ./src/client/Client.cc:9815
#7  0x00007f40e55470bd in Client::_read (this=this@entry=0x1c80cd0, f=0x1d6d8e0, offset=offset@entry=131072, size=size@entry=131072, bl=bl@entry=0x7fffc03d08b0)
    at ./src/client/Client.cc:9701
#8  0x00007f40e5547237 in Client::read (this=0x1c80cd0, fd=10, buf=0x1e5dfa0 "", size=131072, offset=131072) at ./src/client/Client.cc:9601
...
(gdb) info thread
  Id   Target Id         Frame
* 1    Thread 0x7f40e7791740 (LWP 48477) "cephfs-shell" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffc03d06d8)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88 
  2    Thread 0x7f40d79a1700 (LWP 48481) "log" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1c23db4)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88 
  3    Thread 0x7f40d71a0700 (LWP 48482) "io_context_pool" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1b0ccac)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88 
  4    Thread 0x7f40d699f700 (LWP 48483) "io_context_pool" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1b0cca8)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88 
  5    Thread 0x7f40d619e700 (LWP 48484) "msgr-worker-0" 0x00007f40e72afa47 in epoll_wait (epfd=4, events=0x1c51670, maxevents=5000, timeout=30000)
    at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  6    Thread 0x7f40d599d700 (LWP 48485) "msgr-worker-1" 0x00007f40e72afa47 in epoll_wait (epfd=7, events=0x1c600e0, maxevents=5000, timeout=30000)
    at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  7    Thread 0x7f40d519c700 (LWP 48486) "msgr-worker-2" 0x00007f40e72afa47 in epoll_wait (epfd=10, events=0x1cbdf80, maxevents=5000, timeout=30000)
    at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
  8    Thread 0x7f40d499b700 (LWP 48490) "service" 0x00007f40e6f7d065 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f40d499aea0, expected=0,
    futex_word=0x7f40cc009758) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  9    Thread 0x7f40c7fff700 (LWP 48491) "admin_socket" 0x00007f40e72a2cb9 in __GI___poll (fds=0x7f40c7ffee60, nfds=2, timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
  10   Thread 0x7f40c77fe700 (LWP 48492) "ceph_timer" 0x00007f40e6f7d065 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f40c77fde80, expected=0,
    futex_word=0x1c4db64) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  11   Thread 0x7f40c6ffd700 (LWP 48493) "fn_anonymous" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1c4eea8)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88 
  12   Thread 0x7f40c67fc700 (LWP 48494) "safe_timer" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1c80d50)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88 
  13   Thread 0x7f40c5ffb700 (LWP 48495) "fn_anonymous" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
  14   Thread 0x7f40c57fa700 (LWP 48496) "flusher" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
  15   Thread 0x7f40c4ff9700 (LWP 48497) "ms_dispatch" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
  16   Thread 0x7f40bffff700 (LWP 48498) "ms_local" 0x00007f40e6f7cad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1d11468)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  17   Thread 0x7f40bf7fe700 (LWP 48499) "safe_timer" 0x00007f40e6f7d065 in futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7f40bf7fde70, expected=0,
    futex_word=0x1c4fed8) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
  18   Thread 0x7f40beffd700 (LWP 48500) "cephfs-shell" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
(gdb) thread 13
[Switching to thread 13 (Thread 0x7f40c5ffb700 (LWP 48495))]
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
135     ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S: No such file or directory.
(gdb) bt
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007f40e6f79025 in __GI___pthread_mutex_lock (mutex=0x1c80e28) at ../nptl/pthread_mutex_lock.c:80
#2  0x00007f40e54f1510 in __gthread_mutex_lock (__mutex=0x1c80e28) at /usr/include/x86_64-linux-gnu/c++/7/bits/gthr-default.h:748
#3  std::mutex::lock (this=0x1c80e28) at /usr/include/c++/7/bits/std_mutex.h:103
#4  std::scoped_lock<std::mutex>::scoped_lock (__m=..., this=<synthetic pointer>) at /usr/include/c++/7/mutex:610
#5  Client::C_Readahead::finish (this=0x1d79040, r=<optimized out>) at ./src/client/Client.cc:9780
#6  0x00007f40e554bd59 in Context::complete (this=0x1d79040, r=<optimized out>) at ./src/include/Context.h:99
#7  0x00007f40e554bd59 in Context::complete (this=0x1dba7d0, r=<optimized out>) at ./src/include/Context.h:99
#8  0x00007f40e55a6a98 in finish_contexts<std::__cxx11::list<Context*, std::allocator<Context*> > > (cct=0x1be3210, finished=empty std::__cxx11::list, result=result@entry=0)
    at ./src/include/Context.h:209
#9  0x00007f40e559821e in ObjectCacher::bh_read_finish (this=this@entry=0x1c4ebe0, poolid=<optimized out>, oid=..., tid=tid@entry=2, start=131072,
    length=length@entry=131072, bl=..., r=0, trust_enoent=true) at ./src/osdc/ObjectCacher.cc:1000
#10 0x00007f40e55a6dfc in ObjectCacher::C_ReadFinish::finish (this=0x1da40b0, r=0) at ./src/osdc/ObjectCacher.cc:64
#11 0x00007f40e554bd59 in Context::complete (this=0x1da40b0, r=<optimized out>) at ./src/include/Context.h:99
#12 0x00007f40e554eaf5 in C_Lock::finish (this=0x1d83e60, r=<optimized out>) at ./src/include/Context.h:229
#13 0x00007f40e554bd59 in Context::complete (this=0x1d83e60, r=<optimized out>) at ./src/include/Context.h:99
#14 0x00007f40dc68a245 in Finisher::finisher_thread_entry() () from /usr/lib/ceph/libceph-common.so.2
#15 0x00007f40e6f766db in start_thread (arg=0x7f40c5ffb700) at pthread_create.c:463
#16 0x00007f40e72af71f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
````

**Review by batrick — CHANGES_REQUESTED** [review]
Suggest we close this.

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc:9765]: Yeah, right. 
In the followed inode lock patches I have changed the logic about this. 
Will close it now.
Thanks.

**Pull request closed without merging**